jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -327,18 +334,38 @@ default long getLength()
         void demand(Runnable demandCallback);
 
         /**
-         * <p>Fails this content source, possibly failing and discarding accumulated
-         * content chunks that were not yet read.</p>
+         * <p>Fails this content source with a {@link Chunk#isLast() last} {@link Chunk#getFailure() failure chunk},
+         * failing and discarding accumulated content chunks that were not yet read.</p>
          * <p>The failure may be notified to the content reader at a later time, when
-         * the content reader reads a content chunk, via an {@link Chunk.Error} instance.</p>
+         * the content reader reads a content chunk, via a {@link Chunk} instance
+         * with a non null {@link Chunk#getFailure()}.</p>
          * <p>If {@link #read()} has returned a last chunk, this is a no operation.</p>
          * <p>Typical failure: the content being aborted by user code, or idle timeouts.</p>
          * <p>If this method has already been called, then it is a no operation.</p>
          *
          * @param failure the cause of the failure
+         * @see Chunk#getFailure()
          */
         void fail(Throwable failure);
 
+        /**
+         * <p>Fails this content source with a {@link Chunk#getFailure() failure chunk}
+         * that may or not may be {@link Chunk#isLast() last}.
+         * If {@code last} is {@code true}, then the failure is persistent and a call to this method acts
+         * as {@link #fail(Throwable)}. Otherwise the failure is transient and a
+         * {@link Chunk#getFailure() failure chunk} will be {@link #read() read} in order with content chunks,
+         * and subsequent calls to {@link #read() read} may produce other content.</p>
+         * <p>A reader may treat a transient failure as persistent.</p>
+         *
+         * @param failure A failure.
+         * @param last true if the failure is persistent, false if the failure is transient.
+         * @see Chunk#getFailure()
+         */
+        default void fail(Throwable failure, boolean last)
+        {
+            fail(failure);
```

Comment:
I updated the contract to say that a source or its reader can treat a transient failure as permanent.    I.E. it is better to hard fail the source rather than the caller if you don't explicitly support transient.

Reply:
Currently this method is not used, but will be used for idle timeouts from the other PR?